packages/mds-event-processor/src/proc-event.ts

```diff
@@ -0,0 +1,276 @@
+  }
+  const tripsCache = await cache.readTripsTelemetry(`${provider_id}:${device_id}`)
+  const trips = tripsCache || {}
+  // TODO reduce logic
+  if (typeof tripId === 'string') {
+    if (!trips[tripId]) {
+      trips[tripId] = []
+    }
+    trips[tripId].push(tripTelemetry)
+    await cache.writeTripsTelemetry(`${provider_id}:${device_id}`, trips)
+  }
+  return true
+}
+
+async function processTripEvent(deviceState: StateEntry) {
+  /*
+  Add vehicle events of a trip to cache (trips:events):
+
+    Key: 'provider_id:device_id'
+    Value: hash map of tripEvents keyed by trip_id
+
+  */
+  const {
+    timestamp,
+    event_type,
+    event_type_reason,
+    annotation_version,
+    annotation,
+    gps,
+    service_area_id,
+    trip_id,
+    provider_id,
+    device_id
+  } = deviceState
+
+  const tripEvent = {
+    vehicle_type: 'scooter',
+    timestamp,
+    event_type,
+    event_type_reason,
+    annotation_version,
+    annotation,
+    gps,
+    service_area_id
+  } as TripEvent
+
+  // Either append to existing trip or create new entry
+  const tripsCache = await cache.readTripsEvents(`${provider_id}:${device_id}`)
+  const trips = tripsCache || {}
+  // TODO reduce logic
+  if (trip_id) {
+    if (!trips[trip_id]) {
+      trips[trip_id] = []
+    }
+    trips[trip_id].push(tripEvent)
+    // Update trip event cache and stream
+    await cache.writeTripsEvents(`${provider_id}:${device_id}`, trips)
+    // await stream.writeCloudEvent('mds.trip.event', JSON.stringify(tripEvent))
+  }
+  await processTripTelemetry(deviceState)
+}
+
+async function processRaw(type: CE_TYPE, data: InboundEvent & InboundTelemetry) {
+  const { timestamp, device_id, provider_id, recorded } = data as {
+    timestamp: number
+    device_id: string
+    provider_id: string
+    recorded: number
+  }
+  const lastState = await cache.readDeviceState(`${provider_id}:${device_id}`)
+  // Construct state
+  const baseDeviceState = {
+    vehicle_type: 'scooter',
+    type: type.substring(type.lastIndexOf('.') + 1),
+    timestamp,
+    device_id,
+    provider_id,
+    recorded,
+    annotation_version: getAnnotationVersion()
+  } as StateEntry
+
+  switch (baseDeviceState.type) {
+    case 'event': {
+      const { event_type, telemetry, event_type_reason, trip_id, service_area_id } = data as {
+        event_type: VEHICLE_EVENT
+        telemetry: Telemetry
+        event_type_reason: VEHICLE_REASON
+        trip_id: string
+        service_area_id: string
+      }
+      const gps = telemetry?.gps
+      const charge = telemetry?.charge
+      const annotation = gps ? getAnnotationData(gps) : null
+      const deviceState = {
+        ...baseDeviceState,
+        annotation,
+        gps,
+        service_area_id,
+        charge,
+        state: EVENT_STATUS_MAP[event_type],
+        event_type,
+        event_type_reason,
+        trip_id
+      } as StateEntry
+      // Take necessary steps on event trasitions
+      switch (data.event_type) {
```

Comment:
This still strikes me as quite odd. It looks like it's written as though each case should be handled differently, but AFAIK they're all doing exactly the same thing (except default). Check out this article https://www.digitalocean.com/community/tutorials/how-to-use-the-switch-statement-in-javascript#multiple-cases

Alternatively I think this could be rewritten using `if/else`.

Reply:
It's more there if we want to do different things with them in the future